app/controllers/oidc_controller.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require "velum/kubernetes"
+
+# OidcController is used for performing an OpenID Connect auth
+# flow against Dex, allowing Velum to generate a complete kubeconfig
+# file for a user, including their JWT token
+class OidcController < ApplicationController
+  # allow anyone to start this, the auth will happen inside
+  # Dex.
+  skip_before_action :authenticate_user!
+  skip_before_action :redirect_to_setup
+
+  # Skipping the CSRF validation just in case. If we add more methods, then
+  # we'll need to add an "except" or an "only" clause to this.
+  skip_before_action :verify_authenticity_token
+
+  def new_nonce
+    session[:nonce] = SecureRandom.hex(16)
+  end
+
+  def stored_nonce
+    n = session.delete(:nonce)
+    raise "Invalid Session" if n.nil?
+    n
+  end
+
+  def issuer
+    "https://#{Pillar.value(pillar: :apiserver)}:32000"
+  end
+
+  def client_id
+    "caasp-cli"
+  end
+
+  def index
+    lookup_config
+
+    if @apiserver_host.blank?
+      redirect_to root_path,
+                  alert: "Provisioning did not yet finish. Please wait until the cluster is ready."
+    else
+      redirect_to authorization_uri(new_nonce)
+    end
+  end
+
+  def decode_id(id_token)
+    OpenIDConnect::ResponseObject::IdToken.decode(id_token, oidc_config.jwks)
+  end
+
+  def done
+    client = client()
+    client.authorization_code = params[:code]
+
+    access_token = client.access_token!(:basic)
+    id_token = decode_id(access_token.id_token)
+    id_token.verify!(
+      issuer:    issuer,
+      client_id: client_id,
+      nonce:     stored_nonce
+    )
+
+    lookup_config
+
+    @access_token = access_token
+    @id_token = id_token
+
+    response.headers["Content-Disposition"] = "attachment; filename=kubeconfig"
+    render "kubeconfig.erb", layout: false, content_type: "text/yaml"
+  end
+
+  def lookup_config
+    kubeconfig = Velum::Kubernetes.kubeconfig
+    @apiserver_host = kubeconfig.host
+    @ca_crt = kubeconfig.ca_crt
+    @client_crt = kubeconfig.client_crt
+    @client_key = kubeconfig.client_key
+  end
+
+  def authorization_uri(nonce)
+    client = client()
+    client.authorization_uri(
+      response_type: :code,
+      nonce:         nonce,
+      state:         nonce,
+      scope:         [:openid, :profile, :email, :offline_access, :groups].collect(&:to_s)
+    )
+  end
+
+  def oidc_config
+    @oidc_config ||= OpenIDConnect::Discovery::Provider::Config.discover! \
+      "https://#{Pillar.value(pillar: :apiserver)}:32000"
```

Comment:
On my deployment I see that the dashboard container is not able to resolve the hostname I have set to the apiserver. My point here: the `apiserver` pillar is the client reachable name of the `apiserver` machine, but that doesn't mean the admin node will be able to resolve this name.

I think it would be interesting to get the `caasp_fqdn` grain for one machine having the `kube-master` role, and then perform this discovery using this name (the admin node has this information on `/etc/hosts` so it will be able to reach the apiserver with this name, and the apiserver cert contains this name). In my current instance it looks like: `aeda402b236d41ab8476fb9cdefaba3a.infra.caasp.local`.

This would suffice to get this information: `Minion.master.sample.salt.info["caasp_fqdn"]`, so something like this:

```
 @oidc_config ||= OpenIDConnect::Discovery::Provider::Config.discover! \
      "https://#{Minion.master.sample.salt.info["caasp_fqdn"]}:32000"
```

Something similar could be done for the `issuer` method (`Minion.master.sample` could take different masters though, so maybe we could go with `first` for now).

Reply:
That won't work. It must be the same name, or we're going to have certificate issues. Why is your machine different than mine? I'm using caasp-kvm, and it works?

Sounds like we must update the /etc/hosts file on the admin node so that it's always able to find itself by the same name the user uses.

Reply:
> That won't work. It must be the same name, or we're going to have certificate issues. Why is your machine different than mine? I'm using caasp-kvm, and it works?

If we use `first` instead of `sample` it should work. The admin node has the `/etc/hosts` updated automatically by salt but the container has its own ideas about the `/etc/hosts` file.

I'll check this PR first thing in the morning again, not sure why it didn't work for me, even forcing the IP address (hardcoding it) -- that's when I got the `Issuer mismatch`, maybe something wrong on my side 😞

Reply:
As in it should work I mean we should be referring to the same machine, if that would or not work with OpenID I have no idea (I don't know OpenID connect internals well enough).

I confirmed this `caasp_fqdn` name is present on the apiserver certificate, but didn't check further.